[cond] support output sizes mismatch in front end

ydwu4 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #147567
* #147649
* __->__ #147130

This PR finishes https://github.com/pytorch/pytorch/pull/137615 by addressing the TODOs and comments left there.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

ydwu4 committed:
[cond] support output sizes mismatch in front end

[ghstack-poisoned]

ydwu4 committed:
Update on "[cond] support output sizes mismatch in front end"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "[cond] support output sizes mismatch in front end"


This PR finishes https://github.com/pytorch/pytorch/pull/137615 by addressing the TODOs and comments left there.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "[cond] support output sizes mismatch in front end"


This PR finishes https://github.com/pytorch/pytorch/pull/137615 by addressing the TODOs and comments left there.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "[cond] support output sizes mismatch in front end"


This PR finishes https://github.com/pytorch/pytorch/pull/137615 by addressing the TODOs and comments left there.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "[cond] support output sizes mismatch in front end"


This PR finishes https://github.com/pytorch/pytorch/pull/137615 by addressing the TODOs and comments left there.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "[cond] support output sizes mismatch in front end"


This PR finishes https://github.com/pytorch/pytorch/pull/137615 by addressing the TODOs and comments left there.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

zou3519 commented on torch/_higher_order_ops/cond.py:
```diff
@@ -429,30 +379,154 @@ def cond_fake_tensor_mode(mode, pred, true_fn, false_fn, operands):
+
+        Without loss of generality, suppose we have a tenosr with:
+            size [3, 4, 3, 5, 4, 5],
+            stride (1200, 300, 1, 12, 3, 60),
```
I'd recommend testing some tensors with multiple strides of size 1

zou3519 replied:
So concretely I think this PR would benefit from unit testing this _merge_tensors function. 

it would be good to have a single test case where you can specify a bunch of (tensor, tensor) size/stride pairs and make sure those work. For example:
```
data = [
    [([3, 4, 5], [20, 5, 1]), ([5, 4, 5], [20, 5, 1]), "success"],
    [([3, 4, 5], [20, 5, 1]), ([5, 4, 5], [20, 5, 2]), "fail"], 
    ...
]
```

zou3519 commented on torch/_higher_order_ops/cond.py:
```diff
@@ -429,30 +379,154 @@ def cond_fake_tensor_mode(mode, pred, true_fn, false_fn, operands):
+        In each step, it checks whether the current stride is bounded or not and bound old stride
+        to new value. It also generates the next candidate by bounding stride[i] * size[i]
+        to bounded_stride * merged_size[i]. Concretely:
+        1. In 1st iteration, we bound 1 to 1, and set candidates[3] = 3
+        2. In 2nd iteration, we bound 3 to 3, and set candiates[12] = 12
+        3. In 3rd iteration, we bound 12 to 12, and set candidates[60] = 12 * u0
+        4. In 4th iteration, we bound 60 to 12 * u0, and set candidates[300] = 12 * u0 * u0
```
We can only do this assuming the tensor is dense in memory, right? We should error if the tensor isn't dense in memory (is there a check for that somewhere?)

zou3519 replied:
I don't see a check for that

zou3519 commented on torch/_higher_order_ops/cond.py:
```diff
@@ -429,30 +379,154 @@ def cond_fake_tensor_mode(mode, pred, true_fn, false_fn, operands):
+    if a is None or b is None:
+        assert a is None and b is None, (a, b)
+        return None
+    torch._check(
+        a.dtype == b.dtype,
+        lambda: "When merging two branches' output in torch.cond, "
+        f"expected same dtype but got {a.dtype} and {b.dtype}",
+    )
+    torch._check(
+        a.device == b.device,
+        lambda: "When merging two branches' output in torch.cond, "
+        f"expected same device but got {a.device} and {b.device}",
+    )
+    torch._check(
+        a.dim() == b.dim(),
+        lambda: "When merging two branches' output in torch.cond, "
+        f"expected same dim but got {a.dim()} and {b.dim()}",
+    )
```
Are these the only things to be checked? If not it might be nice to have a helper function that easily checks a single property and then we can add more properties to it.

There's also storage_offset, maybe more properties (like is_conj/is_neg, dispatchkeys, etc)

zou3519 commented on torch/_higher_order_ops/cond.py:
```diff
@@ -429,30 +379,154 @@ def cond_fake_tensor_mode(mode, pred, true_fn, false_fn, operands):
+    )
+
+    # NYI
+    assert not a.is_quantized
+    assert not b.is_quantized
```
as I mentioned above there are more properties than just size/stride/dim/device/dtype/is_quantized, we should make an extensible list because we're going to find more properties to throw in

zou3519 commented on torch/_higher_order_ops/cond.py:
```diff
@@ -429,30 +379,154 @@ def cond_fake_tensor_mode(mode, pred, true_fn, false_fn, operands):
+    assert not b.is_quantized
+
+    with mode:
+        return torch.empty_strided(
+            merged_size, a_stride, dtype=a.dtype, device=a.device
+        )
```
Should also check or assert that the inputs are regular Tensors (not tensor subclasses, those should have evaporated by now)

zou3519 reviewed (approved):
LGTM but please read comments

ydwu4 committed:
Update on "[cond] support output sizes mismatch in front end"


This PR finishes https://github.com/pytorch/pytorch/pull/137615 by addressing the TODOs and comments left there.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "[cond] support output sizes mismatch in front end"


This PR finishes https://github.com/pytorch/pytorch/pull/137615 by addressing the TODOs and comments left there.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "[cond] support output sizes mismatch in front end"


This PR finishes https://github.com/pytorch/pytorch/pull/137615 by addressing the TODOs and comments left there.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "[cond] support output sizes mismatch in front end"


This PR finishes https://github.com/pytorch/pytorch/pull/137615 by addressing the TODOs and comments left there.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "[cond] support output sizes mismatch in front end"


This PR finishes https://github.com/pytorch/pytorch/pull/137615 by addressing the TODOs and comments left there.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "[cond] support output sizes mismatch in front end"


This PR finishes https://github.com/pytorch/pytorch/pull/137615 by addressing the TODOs and comments left there.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

ydwu4 commented:
@pytorchbot merge

ydwu4 committed:
Update on "[cond] support output sizes mismatch in front end"


This PR finishes https://github.com/pytorch/pytorch/pull/137615 by addressing the TODOs and comments left there.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]
